File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/ECPrivateKey.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bccrypto.ECDSA_PRIVATE_KEY_SIZE
import com.blockchaincommons.bccrypto.ECDSA_SIGNATURE_SIZE
import com.blockchaincommons.bccrypto.SCHNORR_SIGNATURE_SIZE
import com.blockchaincommons.bccrypto.deriveSigningPrivateKey
import com.blockchaincommons.bccrypto.ecdsaPublicKeyFromPrivateKey
import com.blockchaincommons.bccrypto.ecdsaSign
import com.blockchaincommons.bccrypto.schnorrPublicKeyFromPrivateKey
import com.blockchaincommons.bccrypto.schnorrSign
import com.blockchaincommons.bccrypto.schnorrSignUsing
import com.blockchaincommons.bcrand.RandomNumberGenerator
import com.blockchaincommons.bcrand.SecureRandomNumberGenerator
import com.blockchaincommons.bctags.TAG_EC_KEY
import com.blockchaincommons.bctags.TAG_EC_KEY_V1
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborMap
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * A private key for elliptic curve digital signature algorithms on secp256k1.
 *
 * An [ECPrivateKey] is a 32-byte secret value that can be used to generate its
 * corresponding public key, sign messages using ECDSA, and sign messages using
 * the Schnorr signature scheme (BIP-340).
 *
 * These keys use the secp256k1 curve, the same curve used in Bitcoin and
 * other cryptocurrencies.
 */
class ECPrivateKey private constructor(private val keyData: ByteArray) :
    ECKey,
    ReferenceProvider,
    CborTaggedCodable,
    URCodable {

    init {
        require(keyData.size == KEY_SIZE) {
            "EC private key must be exactly $KEY_SIZE bytes, got ${keyData.size}"
        }
    }

    // -- ECKeyBase --

    override fun data(): ByteArray = keyData.copyOf()

    // -- ECKey --

    /** Derives the corresponding ECDSA compressed public key. */
    override fun publicKey(): ECPublicKey =
        ECPublicKey.fromData(ecdsaPublicKeyFromPrivateKey(keyData))

    /** Derives the Schnorr (x-only) public key from this private key. */
    fun schnorrPublicKey(): SchnorrPublicKey =
        SchnorrPublicKey.fromData(schnorrPublicKeyFromPrivateKey(keyData))

    /**
     * Signs a message using ECDSA.
     *
     * @param message the data to sign
     * @return a 64-byte compact ECDSA signature
     */
    fun ecdsaSign(message: ByteArray): ByteArray {
        val sig = ecdsaSign(keyData, message)
        require(sig.size == ECDSA_SIGNATURE_SIZE)
        return sig
    }

    /**
     * Signs a message using the Schnorr signature scheme (BIP-340).
     *
     * Uses the secure random number generator for nonce generation.
     *
     * @param message the data to sign
     * @return a 64-byte Schnorr signature
     */
    fun schnorrSign(message: ByteArray): ByteArray {
        val sig = schnorrSign(keyData, message)
        require(sig.size == SCHNORR_SIGNATURE_SIZE)
        return sig
    }

    /**
     * Signs a message using the Schnorr signature scheme with a custom RNG.
     *
     * @param message the data to sign
     * @param rng the random number generator to use for nonce generation
     * @return a 64-byte Schnorr signature
     */
    fun schnorrSignUsing(
        message: ByteArray,
        rng: RandomNumberGenerator,
    ): ByteArray {
        val sig = schnorrSignUsing(keyData, message, rng)
        require(sig.size == SCHNORR_SIGNATURE_SIZE)
        return sig
    }

    // -- ReferenceProvider --

    override fun reference(): Reference =
        Reference.fromDigest(Digest.fromImage(taggedCbor().toCborData()))

    // -- CBOR --

    override fun cborTags(): List<Tag> =
        tagsForValues(listOf(TAG_EC_KEY, TAG_EC_KEY_V1))

    /**
     * Encodes as a CBOR map: {2: true, 3: key_bytes}.
     *
     * Key 2 (boolean true) indicates this is a private key.
     * Key 3 is the raw key data as a byte string.
     */
    override fun untaggedCbor(): Cbor {
        val map = CborMap()
        map.insert(Cbor.fromInt(2), Cbor.`true`())
        map.insert(Cbor.fromInt(3), Cbor.fromByteString(keyData))
        return Cbor.fromMap(map)
    }

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ECPrivateKey) return false
        return keyData.contentEquals(other.keyData)
    }

    override fun hashCode(): Int = keyData.contentHashCode()

    // -- toString --

    override fun toString(): String = "ECPrivateKey(${refHexShort()})"

    companion object {
        const val KEY_SIZE: Int = ECDSA_PRIVATE_KEY_SIZE

        /** Creates a new random EC private key using the secure RNG. */
        fun create(): ECPrivateKey {
            val rng = SecureRandomNumberGenerator()
            return createUsing(rng)
        }

        /** Creates a new random EC private key using the given [rng]. */
        fun createUsing(rng: RandomNumberGenerator): ECPrivateKey {
            val data = rng.randomData(KEY_SIZE)
            return ECPrivateKey(data)
        }

        /** Restores an EC private key from exactly [KEY_SIZE] bytes. */
        fun fromData(data: ByteArray): ECPrivateKey {
            if (data.size != KEY_SIZE) {
                throw BcComponentsException.invalidSize(
                    "EC private key", KEY_SIZE, data.size,
                )
            }
            return ECPrivateKey(data.copyOf())
        }

        /**
         * Derives an EC private key from arbitrary key material.
         *
         * Uses HKDF to deterministically derive a valid secp256k1 key.
         */
        fun deriveFromKeyMaterial(keyMaterial: ByteArray): ECPrivateKey =
            fromData(deriveSigningPrivateKey(keyMaterial))

        /**
         * Creates an EC private key from a hexadecimal string.
         *
         * @throws BcComponentsException.InvalidSize if the decoded bytes
         *   are not exactly [KEY_SIZE] bytes
         */
        fun fromHex(hex: String): ECPrivateKey =
            fromData(hex.hexToByteArray())

        /** Decodes an [ECPrivateKey] from untagged CBOR (a map). */
        fun fromUntaggedCbor(cbor: Cbor): ECPrivateKey {
            val map = cbor.tryMap()
            val isPrivate: Boolean = map.get<Int, Boolean>(2)
                ?: throw BcComponentsException.invalidData(
                    "EC key", "missing private key flag (key 2)",
                )
            if (!isPrivate) {
                throw BcComponentsException.invalidData(
                    "EC key", "expected private key (key 2 = true)",
                )
            }
            val bytes: ByteArray = map.extract<Int, ByteArray>(3)
            return fromData(bytes)
        }

        /** Decodes an [ECPrivateKey] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): ECPrivateKey =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_EC_KEY, TAG_EC_KEY_V1)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [ECPrivateKey] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): ECPrivateKey =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_EC_KEY, TAG_EC_KEY_V1)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [ECPrivateKey] from a UR. */
        fun fromUr(ur: UR): ECPrivateKey {
            ur.checkType("eckey")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes an [ECPrivateKey] from a UR string. */
        fun fromUrString(urString: String): ECPrivateKey =
            fromUr(UR.fromUrString(urString))
    }
}

File: python/bc-components/src/bc_components/ec_key/_ec_private_key.py
"""EC private key (32 bytes) on secp256k1."""

from __future__ import annotations

from typing import TYPE_CHECKING

import bc_crypto
from bc_rand import RandomNumberGenerator, SecureRandomNumberGenerator

from .._error import InvalidSizeError
from bc_tags import (
    CBOR,
    Map,
    Tag,
    TAG_EC_KEY,
    TAG_EC_KEY_V1,
    tags_for_values,
)

if TYPE_CHECKING:
    from .._digest import Digest
    from .._reference import Reference

from ._ec_public_key import ECPublicKey
from ._schnorr_public_key import SchnorrPublicKey

ECDSA_PRIVATE_KEY_SIZE: int = bc_crypto.ECDSA_PRIVATE_KEY_SIZE


class ECPrivateKey:
    """A 32-byte private key for secp256k1 elliptic curve algorithms.

    Can be used to:
    - Generate its corresponding public key
    - Sign messages using the ECDSA signature scheme
    - Sign messages using the Schnorr signature scheme (BIP-340)
    """

    KEY_SIZE: int = ECDSA_PRIVATE_KEY_SIZE

    __slots__ = ("_data",)

    def __init__(self, data: bytes) -> None:
        if len(data) != ECDSA_PRIVATE_KEY_SIZE:
            raise InvalidSizeError(
                "EC private key", ECDSA_PRIVATE_KEY_SIZE, len(data)
            )
        self._data = bytes(data)

    @staticmethod
    def generate() -> ECPrivateKey:
        """Create a new random EC private key using a secure RNG."""
        rng = SecureRandomNumberGenerator()
        return ECPrivateKey.generate_using(rng)

    @staticmethod
    def generate_using(rng: RandomNumberGenerator) -> ECPrivateKey:
        """Create a new random EC private key using the given RNG."""
        key_bytes = rng.random_data(ECDSA_PRIVATE_KEY_SIZE)
        return ECPrivateKey.from_data(bytes(key_bytes))

    @staticmethod
    def from_data(data: bytes | bytearray) -> ECPrivateKey:
        """Create from binary data, with size validation."""
        return ECPrivateKey(bytes(data))

    @classmethod
    def from_hex(cls, hex_str: str) -> ECPrivateKey:
        """Create from a hexadecimal string."""
        return cls.from_data(bytes.fromhex(hex_str))

    @staticmethod
    def derive_from_key_material(key_material: bytes | bytearray) -> ECPrivateKey:
        """Derive a private key from the given key material.

        Uses HKDF to deterministically generate a valid private key
        for the secp256k1 curve.
        """
        return ECPrivateKey.from_data(
            bc_crypto.derive_signing_private_key(bytes(key_material))
        )

    @property
    def data(self) -> bytes:
        """Return the key's binary data."""
        return self._data

    def hex(self) -> str:
        """Return the key as a hexadecimal string."""
        return self._data.hex()

    def public_key(self) -> ECPublicKey:
        """Derive the corresponding ECDSA compressed public key."""
        return ECPublicKey.from_data(
            bc_crypto.ecdsa_public_key_from_private_key(self._data)
        )

    def schnorr_public_key(self) -> SchnorrPublicKey:
        """Derive the Schnorr (x-only) public key from this private key."""
        return SchnorrPublicKey.from_data(
            bc_crypto.schnorr_public_key_from_private_key(self._data)
        )

    def ecdsa_sign(self, message: bytes | bytearray) -> bytes:
        """Sign a message using ECDSA. Returns a 64-byte compact signature."""
        return bc_crypto.ecdsa_sign(self._data, bytes(message))

    def schnorr_sign_using(
        self,
        message: bytes | bytearray,
        rng: RandomNumberGenerator,
    ) -> bytes:
        """Sign using Schnorr with a custom RNG. Returns a 64-byte signature."""
        return bc_crypto.schnorr_sign_using(self._data, bytes(message), rng)

    def schnorr_sign(self, message: bytes | bytearray) -> bytes:
        """Sign using Schnorr with a secure RNG. Returns a 64-byte signature."""
        rng = SecureRandomNumberGenerator()
        return self.schnorr_sign_using(message, rng)

    # --- CBOR ---

    @staticmethod
    def cbor_tags() -> list[Tag]:
        return tags_for_values([TAG_EC_KEY, TAG_EC_KEY_V1])

    def untagged_cbor(self) -> CBOR:
        """CBOR map with key 2 = true (private) and key 3 = byte string of key data."""
        m = Map()
        m.insert(2, True)
        m.insert(3, CBOR.from_bytes(self._data))
        return CBOR.from_map(m)

    def tagged_cbor(self) -> CBOR:
        tags = self.cbor_tags()
        cbor = self.untagged_cbor()
        for tag in reversed(tags):
            cbor = CBOR.from_tagged_value(tag, cbor)
        return cbor

    def tagged_cbor_data(self) -> bytes:
        return self.tagged_cbor().to_cbor_data()

    @staticmethod
    def from_untagged_cbor(cbor: CBOR) -> ECPrivateKey:
        m = cbor.try_map()
        key_data = m.extract(3).try_byte_string()
        return ECPrivateKey.from_data(key_data)

    @staticmethod
    def from_tagged_cbor(cbor: CBOR) -> ECPrivateKey:
        tags = ECPrivateKey.cbor_tags()
        inner = cbor
        for tag in tags:
            inner = inner.try_expected_tagged_value(tag.value)
        return ECPrivateKey.from_untagged_cbor(inner)

    @staticmethod
    def from_tagged_cbor_data(data: bytes) -> ECPrivateKey:
        return ECPrivateKey.from_tagged_cbor(CBOR.from_data(data))

    # --- Reference ---

    def reference(self) -> Reference:
        from .._digest import Digest
        from .._reference import Reference

        return Reference.from_digest(
            Digest.from_image(self.tagged_cbor_data())
        )

    def ref_hex_short(self) -> str:
        return self.reference().ref_hex_short()

    def __eq__(self, other: object) -> bool:
        if isinstance(other, ECPrivateKey):
            return self._data == other._data
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._data)

    def __repr__(self) -> str:
        return f"ECPrivateKey({self.hex()})"

    def __str__(self) -> str:
        return f"ECPrivateKey({self.ref_hex_short()})"
